File: tests/unit/python/execution_tree/primitives/make_list.py
#  Copyright (c) 2018 Hartmut Kaiser
#
#  Distributed under the Boost Software License, Version 1.0. (See accompanying
#  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)

# flake8: noqa

from phylanx import Phylanx, PhylanxSession

PhylanxSession(1)


@Phylanx
def test_make_list_empty():
    return make_list()


@Phylanx
def test_make_list_one():
    return make_list(42)


@Phylanx
def test_make_list_literals():
    return make_list(1, 2, 3, 4)


@Phylanx
def test_make_list():
    a = 1
    b = 2
    c = 3
    return make_list(a, b, c)


@Phylanx
def test_make_list2():
    a = 1
    b = 2
    c = 3
    return make_list(a, make_list(a, b, c), c)


@Phylanx
def test_make_list_arg(l):
    a = 1
    c = 3
    return make_list(a, l, c)


assert test_make_list_empty() == []
assert test_make_list_one() == [42]
assert test_make_list_literals() == [1, 2, 3, 4]
assert test_make_list() == [1, 2, 3]
assert test_make_list2() == [1, [1, 2, 3], 3]
assert test_make_list_arg([1, 2, 3]) == [1, [1, 2, 3], 3]

File: src/plugins/listops/make_list.cpp
//  Copyright (c) 2018 Hartmut Kaiser
//
//  Distributed under the Boost Software License, Version 1.0. (See accompanying
//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)

#include <phylanx/config.hpp>
#include <phylanx/plugins/listops/make_list.hpp>

#include <hpx/include/lcos.hpp>
#include <hpx/include/naming.hpp>
#include <hpx/include/util.hpp>
#include <hpx/throw_exception.hpp>

#include <cstddef>
#include <memory>
#include <numeric>
#include <string>
#include <utility>
#include <vector>

#include <blaze/Math.h>

///////////////////////////////////////////////////////////////////////////////
namespace phylanx { namespace execution_tree { namespace primitives
{
    ///////////////////////////////////////////////////////////////////////////
    match_pattern_type const make_list::match_data =
    {
        hpx::util::make_tuple("make_list",
            std::vector<std::string>{"make_list(__1)", "'(__1)"},
            &create_make_list, &create_primitive<make_list>)
    };

    ///////////////////////////////////////////////////////////////////////////
    make_list::make_list(
            std::vector<primitive_argument_type>&& operands,
            std::string const& name, std::string const& codename)
      : primitive_component_base(std::move(operands), name, codename)
    {}

    ///////////////////////////////////////////////////////////////////////////
    hpx::future<primitive_argument_type> make_list::eval(
        std::vector<primitive_argument_type> const& operands,
        std::vector<primitive_argument_type> const& args) const
    {
        auto this_ = this->shared_from_this();
        return hpx::dataflow(hpx::launch::sync, hpx::util::unwrapping(
            [this_](std::vector<primitive_argument_type> && args)
            ->  primitive_argument_type
            {
                return primitive_argument_type{std::move(args)};
            }),
            detail::map_operands(
                operands, functional::value_operand{}, args,
                name_, codename_));
    }

    hpx::future<primitive_argument_type> make_list::eval(
        std::vector<primitive_argument_type> const& args) const
    {
        if (operands_.empty())
        {
            return eval(args, noargs);
        }
        return eval(operands_, args);
    }
}}}
